### PR TITLE
Rudimentary support for iterative trace

### PR DESCRIPTION
**Pull request by paul-pearce**
Provides intermediate queries and results involved for iterative lookups as the .data.trace field (list). Sample below

```json
{
  "data": {
    "trace": [
      {
        "cached": false,
        "layer": ".",
        "depth": 1,
        "name_server": "202.12.27.33:53",
        "name": "www.example.com",
        "class": 1,
        "type": 1,
        "results": {
          "flags": {
            "error_code": 0,
            "response": true,
            "opcode": 0,
            "authoritative": false,
            "truncated": false,
            "recursion_desired": false,
            "recursion_available": false,
            "authenticated": false,
            "checking_disabled": false
          },
          "protocol": "udp",
          "authorities": [
            {
              "answer": "l.gtld-servers.net",
              "name": "com",
              "class": "IN",
              "type": "NS",
              "ttl": 172800
            },
            {
              "answer": "k.gtld-servers.net",
              "name": "com",
              "class": "IN",
              "type": "NS",
              "ttl": 172800
            },
            {
              "answer": "c.gtld-servers.net",
              "name": "com",
              "class": "IN",
              "type": "NS",
              "ttl": 172800
            },
            {
              "answer": "d.gtld-servers.net",
              "name": "com",
              "class": "IN",
              "type": "NS",
              "ttl": 172800
            },
            {
              "answer": "m.gtld-servers.net",
              "name": "com",
              "class": "IN",
              "type": "NS",
              "ttl": 172800
            },
            {
              "answer": "g.gtld-servers.net",
              "name": "com",
              "class": "IN",
              "type": "NS",
              "ttl": 172800
            },
            {
              "answer": "h.gtld-servers.net",
              "name": "com",
              "class": "IN",
              "type": "NS",
              "ttl": 172800
            },
            {
              "answer": "f.gtld-servers.net",
              "name": "com",
              "class": "IN",
              "type": "NS",
              "ttl": 172800
            },
            {
              "answer": "a.gtld-servers.net",
              "name": "com",
              "class": "IN",
              "type": "NS",
              "ttl": 172800
            },
            {
              "answer": "b.gtld-servers.net",
              "name": "com",
              "class": "IN",
              "type": "NS",
              "ttl": 172800
            },
            {
              "answer": "e.gtld-servers.net",
              "name": "com",
              "class": "IN",
              "type": "NS",
              "ttl": 172800
            },
            {
              "answer": "i.gtld-servers.net",
              "name": "com",
              "class": "IN",
              "type": "NS",
              "ttl": 172800
            },
            {
              "answer": "j.gtld-servers.net",
              "name": "com",
              "class": "IN",
              "type": "NS",
              "ttl": 172800
            }
          ],
          "additionals": [
            {
              "answer": "192.5.6.30",
              "name": "a.gtld-servers.net",
              "class": "IN",
              "type": "A",
              "ttl": 172800
            },
            {
              "answer": "192.33.14.30",
              "name": "b.gtld-servers.net",
              "class": "IN",
              "type": "A",
              "ttl": 172800
            },
            {
              "answer": "192.26.92.30",
              "name": "c.gtld-servers.net",
              "class": "IN",
              "type": "A",
              "ttl": 172800
            },
            {
              "answer": "192.31.80.30",
              "name": "d.gtld-servers.net",
              "class": "IN",
              "type": "A",
              "ttl": 172800
            },
            {
              "answer": "192.12.94.30",
              "name": "e.gtld-servers.net",
              "class": "IN",
              "type": "A",
              "ttl": 172800
            },
            {
              "answer": "192.35.51.30",
              "name": "f.gtld-servers.net",
              "class": "IN",
              "type": "A",
              "ttl": 172800
            },
            {
              "answer": "192.42.93.30",
              "name": "g.gtld-servers.net",
              "class": "IN",
              "type": "A",
              "ttl": 172800
            },
            {
              "answer": "192.54.112.30",
              "name": "h.gtld-servers.net",
              "class": "IN",
              "type": "A",
              "ttl": 172800
            },
            {
              "answer": "192.43.172.30",
              "name": "i.gtld-servers.net",
              "class": "IN",
              "type": "A",
              "ttl": 172800
            },
            {
              "answer": "192.48.79.30",
              "name": "j.gtld-servers.net",
              "class": "IN",
              "type": "A",
              "ttl": 172800
            },
            {
              "answer": "192.52.178.30",
              "name": "k.gtld-servers.net",
              "class": "IN",
              "type": "A",
              "ttl": 172800
            },
            {
              "answer": "192.41.162.30",
              "name": "l.gtld-servers.net",
              "class": "IN",
              "type": "A",
              "ttl": 172800
            },
            {
              "answer": "192.55.83.30",
              "name": "m.gtld-servers.net",
              "class": "IN",
              "type": "A",
              "ttl": 172800
            },
            {
              "answer": "2001:503:a83e::2:30",
              "name": "a.gtld-servers.net",
              "class": "IN",
              "type": "AAAA",
              "ttl": 172800
            }
          ],
          "answers": []
        }
      },
      {
        "cached": false,
        "layer": "com",
        "depth": 2,
        "name_server": "192.41.162.30:53",
        "name": "www.example.com",
        "class": 1,
        "type": 1,
        "results": {
          "flags": {
            "error_code": 0,
            "response": true,
            "opcode": 0,
            "authoritative": false,
            "truncated": false,
            "recursion_desired": false,
            "recursion_available": false,
            "authenticated": false,
            "checking_disabled": false
          },
          "protocol": "udp",
          "authorities": [
            {
              "answer": "a.iana-servers.net",
              "name": "example.com",
              "class": "IN",
              "type": "NS",
              "ttl": 172800
            },
            {
              "answer": "b.iana-servers.net",
              "name": "example.com",
              "class": "IN",
              "type": "NS",
              "ttl": 172800
            }
          ],
          "additionals": [
            {
              "answer": "199.43.135.53",
              "name": "a.iana-servers.net",
              "class": "IN",
              "type": "A",
              "ttl": 172800
            },
            {
              "answer": "2001:500:8f::53",
              "name": "a.iana-servers.net",
              "class": "IN",
              "type": "AAAA",
              "ttl": 172800
            },
            {
              "answer": "199.43.133.53",
              "name": "b.iana-servers.net",
              "class": "IN",
              "type": "A",
              "ttl": 172800
            },
            {
              "answer": "2001:500:8d::53",
              "name": "b.iana-servers.net",
              "class": "IN",
              "type": "AAAA",
              "ttl": 172800
            }
          ],
          "answers": []
        }
      },
      {
        "cached": false,
        "layer": "example.com",
        "depth": 3,
        "name_server": "199.43.135.53:53",
        "name": "www.example.com",
        "class": 1,
        "type": 1,
        "results": {
          "flags": {
            "error_code": 0,
            "response": true,
            "opcode": 0,
            "authoritative": true,
            "truncated": false,
            "recursion_desired": false,
            "recursion_available": false,
            "authenticated": false,
            "checking_disabled": false
          },
          "protocol": "udp",
          "authorities": [
            {
              "answer": "a.iana-servers.net",
              "name": "example.com",
              "class": "IN",
              "type": "NS",
              "ttl": 86400
            },
            {
              "answer": "b.iana-servers.net",
              "name": "example.com",
              "class": "IN",
              "type": "NS",
              "ttl": 86400
            }
          ],
          "additionals": [],
          "answers": [
            {
              "answer": "93.184.216.34",
              "name": "www.example.com",
              "class": "IN",
              "type": "A",
              "ttl": 86400
            }
          ]
        }
      }
    ],
    "flags": {
      "error_code": 0,
      "response": true,
      "opcode": 0,
      "authoritative": true,
      "truncated": false,
      "recursion_desired": false,
      "recursion_available": false,
      "authenticated": false,
      "checking_disabled": false
    },
    "protocol": "udp",
    "authorities": [],
    "additionals": [],
    "answers": [
      {
        "answer": "93.184.216.34",
        "name": "www.example.com",
        "class": "IN",
        "type": "A",
        "ttl": 86400
      }
    ]
  },
  "timestamp": "2017-08-20T16:48:54-07:00",
  "status": "NOERROR",
  "class": "IN",
  "name": "www.example.com"
}
```

### CODE REVIEW

File: modules/miekg/miekg.go [hunk 1]
```diff
@@ -510,17 +524,19 @@ func (s *Lookup) retryingLookup(dnsType uint16, dnsClass uint16, name string, na
 	panic("loop must return")
 }
 
-func (s *Lookup) cachedRetryingLookup(dnsType uint16, dnsClass uint16, name string, nameServer string, layer string, depth int) (Result, zdns.Status, error) {
+func (s *Lookup) cachedRetryingLookup(dnsType uint16, dnsClass uint16, name string, nameServer string, layer string, depth int) (Result, zdns.Status, error, bool) {
```
Comment (zakird) [inline modules/miekg/miekg.go:527]: It'd probably be cleaner to use a typedef for this result.

File: modules/miekg/miekg.go [hunk 2]
```diff
@@ -710,22 +739,25 @@ func (s *Lookup) iterativeLookup(dnsType uint16, dnsClass uint16, name string, n
...
+		return result, zdns.STATUS_ERROR, errors.New("NOERROR record without any answers or authorities"), trace
 	}
 }
 
 func (s *Lookup) DoMiekgLookup(name string) (interface{}, zdns.Status, error) {
 	if s.Factory.IterativeResolution {
 		s.VerboseLog(0, "MIEKG-IN: iterative lookup for ", name, " (", s.DNSType, ")")
 		s.IterativeStop = time.Now().Add(time.Duration(s.Factory.IterativeTimeout))
-		result, status, err := s.iterativeLookup(s.DNSType, s.DNSClass, name, s.NameServer, 1, ".")
+		result, status, err, trace := s.iterativeLookup(s.DNSType, s.DNSClass, name, s.NameServer, 1, ".", make([]Trace, 0))
```
Comment (zakird) [inline modules/miekg/miekg.go]: Trace seems like "data" and should probably come before error.

**Review by zakird — APPROVED** [review]
Overall, looks good to me. A couple of nitpick comments inline.

**Comment by zakird** [comment]
@paul-pearce is this ready to go? Not sure based on your message last night.

**Comment by paul-pearce** [comment]
Closing PR as I'm in the processing of doing a major overhaul in how I handle this, and I think it needs a whole new review.

**Pull request closed without merging**